Block-based template parts for Classic themes

Mamaduka opened the pull request:
## What?
See #37943

PR allows Classic / Hybrid themes to use block-based template parts without using complete block-based templates. These can be rendered in PHP templates using the [`block_template_part`](https://developer.wordpress.org/reference/functions/block_template_part/) method.

## Why?

The feature should make it easier to transition between Classic, Hybrid, and Block themes.

## How?

PR alters Site Editor to allow only listing and editing template parts. It can be enabled for Classic themes using new `add_theme_support( 'block-template-parts' )` theme feature (_not a final name_). Once the feature is enabled new "Template Parts" menu is visible under "Appearance," which displays a list of template parts.

The changes in `template-parts-screen.php` will be absorbed by `site-editor.php`. Inline comments have notes about migration paths.

## Open Questions

* Should we enable alignments at the root level? These are [disabled by default](https://github.com/WordPress/gutenberg/pull/30079) for Site Editor, so root level blocks cannot be aligned when editing template parts.
* ~~Do we want to display the Global Styles sidebar if a theme doesn't have a `theme.json` file?~~
* ~~Do we want to allow users to add new template parts via UI?~~

## Testing Instructions
1. Clone a testing TT1 child theme (requires TT1 parent) - https://github.com/Mamaduka/block-fragments
2. Access the new "Template Parts" menu.
3. Edit the footer template and confirm changes are visible on the front end. Currently, the theme supports only the footer template.
4. Experiment 🧪

## Screenshots or screencast <!-- if applicable -->

https://user-images.githubusercontent.com/240569/181210420-75e2e0f6-9c94-415b-b1a2-58dbe8ec9ba7.mp4

Mamaduka committed:
WIP: Block-based template parts

Mamaduka committed:
Only display template parts menu

Mamaduka commented on packages/edit-site/src/components/navigation-sidebar/navigation-panel/index.js:
```diff
@@ -40,22 +40,23 @@ function NavLink( { params, replace, ...props } ) {
-				homeTemplate: getSettings().__unstableHomeTemplate,
+				homeTemplate: __unstableHomeTemplate,
 				isNavigationOpen: isNavigationOpened(),
+				isTemplatePartsMode: supportsTemplatePartsMode,
```
I need a better name for this. Maybe `templatePartsEnabled`?

Mamaduka committed:
pinking shears

fabiankaegy commented:
This is already looking ✨ fabulous ✨ Thank you for working on it :) 

As far as your open questions go here are my thoughts:
>Do we want to display the Global Styles sidebar if a theme doesn't have a theme.json file?

I'm not even sure whether we want to enable the Global Styles Sidebar at all, even for themes that have `theme.json` files. For me, editing template parts that are coming from the theme has nothing to do with editing global style options. The global styles editor is currently not available for hybrid themes and I don't think that this PR should change that. But of course open for discussion here :) 

>Should we enable alignments at the root level? These are https://github.com/WordPress/gutenberg/pull/30079 for Site Editor, so root level blocks cannot be aligned when editing template parts.

I'm torn on this one. I think that the alignment of the outermost element most likely should be determined by the theme that creates the template part and therefore that we shouldn't have the alignment options on the top level items.

>Do we want to allow users to add new template parts via UI?

I think we also discussed this in the actual issue for this feature and came to the conclusion that users should not be able to create new template parts in the editor since these templates would not get rendered anywhere. Which would be a more confusing experience. See https://github.com/WordPress/gutenberg/issues/37943#issuecomment-1142544238

Mamaduka commented:
I also want to mention this issue, which also touches on the problem of missing alignments in template parts focus mode - https://github.com/WordPress/gutenberg/issues/34010. Cc. @richtabor.

carolinan commented:
I agree with the above, I don't think that the global styles or adding new template parts should be enabled: Or not until there has been more discussions and exploring around it.

fabiankaegy commented:
I also started testing this out locally on a project and ran into this issue: 
![CleanShot 2022-07-27 at 19 55 42@2x](https://user-images.githubusercontent.com/20684594/181340321-9154f645-ad2b-474c-85c2-530d1e1fdf9d.png)

```
The above error occurred in the <Ur> component:

Ur@http://localhost:8888/wp-content/plugins/gutenberg/build/edit-site/index.min.js?ver=22e1ee72d657210add5e:41:32255
Tl@http://localhost:8888/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=1d7ce6dc82d711ff5290:7:10074
xl@http://localhost:8888/wp-content/plugins/gutenberg/build/edit-site/index.min.js?ver=22e1ee72d657210add5e:35:11933
hn@http://localhost:8888/wp-content/plugins/gutenberg/build/core-data/index.min.js?ver=2088be197fa80353851d:1:44423
hn@http://localhost:8888/wp-content/plugins/gutenberg/build/core-data/index.min.js?ver=2088be197fa80353851d:1:44423
div
x@http://localhost:8888/wp-content/plugins/gutenberg/build/keyboard-shortcuts/index.min.js?ver=9a77e75c0efd974c1503:1:3978
Yr@http://localhost:8888/wp-content/plugins/gutenberg/build/edit-site/index.min.js?ver=22e1ee72d657210add5e:57:71
Ne@http://localhost:8888/wp-content/plugins/gutenberg/build/edit-site/index.min.js?ver=22e1ee72d657210add5e:17:6302
oo@http://localhost:8888/wp-content/plugins/gutenberg/build/components/index.min.js?ver=4d085562c1dd13f7301c:1:267223
io@http://localhost:8888/wp-content/plugins/gutenberg/build/components/index.min.js?ver=4d085562c1dd13f7301c:1:268072
lo@http://localhost:8888/wp-content/plugins/gutenberg/build/components/index.min.js?ver=4d085562c1dd13f7301c:1:269921
Lo@http://localhost:8888/wp-content/plugins/gutenberg/build/edit-site/index.min.js?ver=22e1ee72d657210add5e:74:485

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://reactjs.org/link/error-boundaries to learn more about error boundaries.
```

The exact setup here uses a Custom theme with a custom empty template part added. Loading the template list works and it pulls in the template parts. Just the editor itself crashes. 

The template part editor also still allows me to create new templates (and create custom-created ones). (the editor also crashes with the same issue)

carlomanf commented:
> It can be enabled for Classic themes using new `add_theme_support( 'block-template-parts' )` theme feature

This part concerns me because the classification of themes is already complicated enough before introducing yet another subcategory. If a theme has a `parts` subfolder with valid files inside, then it communicates that it supports this feature and shouldn't need to add a line in `functions.php` as well.

The use of a theme support like this also raises some questions about borderline cases:

- What if a theme already includes block template parts but doesn't add the theme support?
- What if a theme adds the theme support but doesn't include any block template parts?
- Would it be possible to support `'block-templates'` (existing theme support) but not `'block-template-parts'`?

> > Do we want to allow users to add new template parts via UI?
> 
> I think we also discussed this in the actual issue for this feature and came to the conclusion that users should not be able to create new template parts in the editor since these templates would not get rendered anywhere. Which would be a more confusing experience. See [#37943 (comment)](https://github.com/WordPress/gutenberg/issues/37943#issuecomment-1142544238)

> I agree with the above, I don't think that the global styles or adding new template parts should be enabled: Or not until there has been more discussions and exploring around it.

While the reasoning here makes sense, it creates a notable disparity between the behaviour of the existing `'block-templates'` and the proposed `'block-template-parts'` theme supports. `'block-templates'` is intended for when the theme actually provides no block templates of its own but is allowing users to create new ones. `'block-template-parts'` is being proposed in the opposite way, for when the theme is providing its own block template parts but is not allowing users to create new ones. That could be equally confusing for theme authors.

fabiankaegy commented:
@carlomanf thanks for your detailed notes :) I can see that disparity. However, I think the main difference to consider here, is that templates have the template picker which also is the main entry point for users to create new templates. So you always know where the template you are creating is going to be used. This is not the case for these template parts as themes manually render these parts within their traditional templates.

carlomanf commented:
@fabiankaegy I'm not saying the behaviour should be re-aligned with block templates. The reasoning you put forward makes sense, as I noted. I'm saying that it would be very easy for theme authors to get confused and wrongly think that the two theme supports work in the same way, when they don't and can't.

Mamaduka commented:
Thanks for the feedback, @carlomanf. The theme support name or the method of how we enable the feature isn't final, and I'm open to suggestions here.

I decided to go with the support flag because scanning directories for template parts can sometimes yield false results.

fabiankaegy commented:
> @fabiankaegy I'm not saying the behaviour should be re-aligned with block templates. The reasoning you put forward makes sense, as I noted. I'm saying that it would be very easy for theme authors to get confused and wrongly think that the two theme supports work in the same way, when they don't and can't.

Thanks for that context. I miss read the original note :)

carlomanf commented:
@Mamaduka There were a few different suggestions already back in #37943. One suggestion was to check for the presence of the `templateParts` key in `theme.json`, which might be promising.

fabiankaegy commented on lib/compat/wordpress-6.1/template-parts-screen.php:
```diff
@@ -0,0 +1,206 @@
+		( ! is_array( $editor_styles ) || count( $editor_styles ) === 0 )
+	) {
+		wp_enqueue_style( 'wp-block-library-theme' );
+	}
```
@Mamaduka I think we may be missing the `enqueue_block_editor_assets` action here. See https://github.com/WordPress/wordpress-develop/blob/52f51f5cdff49e04805233d56560abbbe1813e35/src/wp-admin/site-editor.php#L141

```suggestion
	}
	
	/** This action is documented in wp-admin/edit-form-blocks.php */
	do_action( 'enqueue_block_editor_assets' );
```

Mamaduka replied:
Nice catch, @fabiankaegy!

fabiankaegy commented:
I've been doing some more testing here:

I think we should remove the option to allow editors to create a new template part inline via the blocks context menu:
<img width="280" alt="CleanShot 2022-07-28 at 08 49 05@2x" src="https://user-images.githubusercontent.com/20684594/181439535-25084854-6c67-4277-88dd-73e0dd0bf1f7.png">


As noted before I think both the "Navigation" and the "Global Styles" panels should not be available in the template part editor
<img width="353" alt="CleanShot 2022-07-28 at 08 51 24@2x" src="https://user-images.githubusercontent.com/20684594/181439884-5737bb76-9b24-4129-be17-4518f2920e65.png">

Additionally, one thing that I think would be a really valuable feature that might be out of scope for this here is allowing template parts to be template locked. What I mean by that is allowing a theme to create a template part that has for example a group block at the root of it that is set to align full. Editors should not be able to remove the group block (can already be done with the `{"lock":{"remove": true}}` option.) But editors should also not be able to add any other blocks to the root of the template part. Everything should be contained within the container.

fabiankaegy commented:
Also just noting that the `title` meta tag is currently showing as `Editor (beta)` not sure whether we want to update that to `Template Part Editor`

Mamaduka committed:
Add enqueue_block_editor_assets

Co-authored-by: Fabian Kägy <mail@fabian-kaegy.de>

Mamaduka committed:
Remove whitespace

Mamaduka committed:
Hide Global Styles

Mamaduka committed:
Disable template creation

Mamaduka committed:
Disable make template part action for blocks

Mamaduka commented:
Pushed the following changes:

* Global Styles are now disabled.
* The "Add Template Part" button isn't rendered on the template parts list page.
* The "Make template part" action is also disabled for blocks.

@carlomanf, I'm unsure if we want to make the `theme.json` requirement to enable this feature. The Classic themes have been using theme support for years to opt-in functionality.

@fabiankaegy, it might be better to handle locking enhancements separately. Maybe something like we did for Group and Columns blocks #26128.

fabiankaegy commented:
@Mamaduka I've created a new issue for the ability to "lock" the root of the template https://github.com/WordPress/gutenberg/issues/42775

carlomanf commented:
@Mamaduka What do you think about adding the following to this patch, on an appropriate action hook?

```
if ( current_theme_supports( 'block-templates' ) ) {
    add_theme_support( 'block-template-parts' );
}
```

Is it actually possible for a theme to support block templates without supporting block template parts? If not, then this would make sense for a few reasons:

- If a theme explicitly calls `add_theme_support( 'block-templates' )` only, then its users should not miss out on this feature
- If a plugin or similar calls `current_theme_supports( 'block-template-parts' )`, then it would be strange for it to return `false`
- It would improve the behavioural parity between the two theme supports.

Soean commented on lib/compat/wordpress-6.1/template-parts-screen.php:
```diff
@@ -0,0 +1,209 @@
+ *
+ * @return void
+ */
+function getunberg_register_template_parts_theme_feature() {
```
```suggestion
function gutenberg_register_template_parts_theme_feature() {
```

Soean commented on lib/compat/wordpress-6.1/template-parts-screen.php:
```diff
@@ -0,0 +1,209 @@
+		)
+	);
+}
+add_action( 'setup_theme', 'getunberg_register_template_parts_theme_feature', 5 );
```
```suggestion
add_action( 'setup_theme', 'gutenberg_register_template_parts_theme_feature', 5 );
```

Soean commented on lib/compat/wordpress-6.1/template-parts-screen.php:
```diff
@@ -0,0 +1,209 @@
+	wp_add_inline_script(
+		'wp-blocks',
+		sprintf( 'wp.blocks.setCategories( %s );', wp_json_encode( get_block_categories( $block_editor_context ) ) ),
+		'after'
```
This is the default value, we don't need it here.

```suggestion
```

Mamaduka replied:
I copied this from `site-editor.php,` so we might want to check the editor pages in core and remove them.

Soean reviewed (commented):
Thanks, this is a really cool feature.
I was able to test it, just found a minor typo.

Mamaduka committed:
Update lib/compat/wordpress-6.1/template-parts-screen.php

Co-authored-by: Sören Wrede <soerenwrede@gmail.com>

Mamaduka committed:
Update lib/compat/wordpress-6.1/template-parts-screen.php

Co-authored-by: Sören Wrede <soerenwrede@gmail.com>

Mamaduka committed:
Update lib/compat/wordpress-6.1/template-parts-screen.php

Co-authored-by: Sören Wrede <soerenwrede@gmail.com>

Mamaduka committed:
Remove trailing comma

carolinan commented:
> @Mamaduka What do you think about adding the following to this patch, on an appropriate action hook?
> 
> ```
> if ( current_theme_supports( 'block-templates' ) ) {
>     add_theme_support( 'block-template-parts' );
> }
> ```
> 
> Is it actually possible for a theme to support block templates without supporting block template parts? If not, then this would make sense for a few reasons:
> 
> * If a theme explicitly calls `add_theme_support( 'block-templates' )` only, then its users should not miss out on this feature
> * If a plugin or similar calls `current_theme_supports( 'block-template-parts' )`, then it would be strange for it to return `false`
> * It would improve the behavioural parity between the two theme supports.

I think that they need to be separate, because we don't know all the possible use cases that developers may want to use.
It is easier for developers to add both supports if they want the theme to support both, than to separate them if this is enforced?

carlomanf commented:
But isn't it the case that users can create template parts through the template editor? If the intention is for them to be separate, then template parts should be hidden from the block inserter. In any case, the theme support should still be automatically added for pure block themes.

fabiankaegy commented:
Currently, if you use the template editor today you can only create and edit block template parts by using the `template-part` block. Because of that, I agree with @carlomanf that it would probably be a nice enhancement for everybody that already allows for block-templates to get the template part editor automatically. 

We should also make the option to create new template parts depending on whether or not the theme supports block templates. Since editors are already able to create new template parts using the template editor it would be great if they could also do the same in the template part editor. But only if the theme supports block templates since without it there is no way for editors to actually use the template parts they created without going into the code to call them manually. 

**The current state as it is would be:**

| Theme Supports                                    | Can edit Template Parts _(in isolation)_ | Can add new Template Parts | Can edit Templates |
| --------------------------------------|-------------------------|------------------------------|--------------------|
| Traditional Theme _(no theme.json)_                  | ❌                                     | ❌                                            | ❌                             |
| `add_theme_support( 'block-template-parts' )`| ✅                                       | ❌                                              | ❌                             |
| `add_theme_support( 'block-templates' )` | ❌                                     | ✅                                             | ✅                             |
| `add_theme_support( 'block-template-parts' )` & `add_theme_support( 'block-templates' )`| ✅                                      | ✅                                             | ✅                             |
| Block Based Theme                       | ✅ _(in the actual site editor)_                                     | ✅                                              | ✅                             |

**I think the ideal state would be:**

| Theme Supports                                    | Can edit Template Parts _(in isolation)_ | Can add new Template Parts | Can edit Templates |
| --------------------------------------|-------------------------|------------------------------|--------------------|
| Traditional Theme _(no theme.json)_                  | ❌                                     | ❌                                            | ❌                             |
| `add_theme_support( 'block-template-parts' )`| ✅                                       | ❌                                              | ❌                             |
| `add_theme_support( 'block-templates' )` | ✅                                      | ✅                                             | ✅                             |
| Block Based Theme                       | ✅ _(in the actual site editor)_                                     | ✅                                              | ✅                             |

Mamaduka commented:
> In any case, the theme support should still be automatically added for pure block themes.

@carlomanf, what will be the use case for this? The whole Site Editor is available for block themes.

I'm still lean towards having `block-templates` and `block-template-parts` enabled separately.

The support flag name isn't final, just a placeholder I decided to use in the initial version. If you think that name similarities would be confusing for theme developers, then we should try and come up with a better name.

carlomanf commented:
> > In any case, the theme support should still be automatically added for pure block themes.
> 
> @carlomanf, what will be the use case for this? The whole Site Editor is available for block themes.

Plugins and/or core needing to do things on the condition that the theme supports template parts would be able to simply and more intuitively check `current_theme_supports( 'block-template-parts' )` rather than the more complex and less intuitive `current_theme_supports( 'block-template-parts' ) || wp_is_block_theme()`. This follows the pattern of the existing block templates theme support, which is automatically added for pure block themes.

> I'm still lean towards having `block-templates` and `block-template-parts` enabled separately.

I'm OK with this as well, but I think the Template Part block should be hidden from the block inserter if `'block-template-parts'` is not enabled. It makes no sense for the management screen to be hidden but the block inserter available. It also should be noted that the Template Part block was not actually part of the original template editor in 5.8, but appeared in 5.9 with no separate opt-in mechanism. This would be a chance to rectify that.

georgestephanis commented:
@Mamaduka Just a question on the current status quo of this -- if I was using a wholly classic theme that doesn't even have a `theme.json` -- a stock `twentytwenty` for example -- is this something that a plugin would be able to opt in to defining a "TippyTop" template part that could be injected on the `wp_body_open` action, without any modifications in the theme?  If so, could a barebones example for that be put out as a pattern to follow?

That's precisely a pain point I've been experiencing lately, wanting to backfill and use some of the FSE tooling in classic themes -- but coming at it from a plugin direction instead of modifying the theme.

fabiankaegy commented on packages/edit-site/src/components/list/header.js:
```diff
@@ -9,10 +9,17 @@ import { __experimentalHeading as Heading } from '@wordpress/components';
+			const settings = select( editSiteStore ).getSettings();
+			return {
+				postType: select( coreStore ).getPostType( templateType ),
+				canCreate: ! settings?.supportsTemplatePartsMode,
```
I would change this to be dependent on whether the theme supports the Template Editing mode. 

So it would not be allowed if the Template Part Editing feature is enabled on its own. But once the Template Editor also gets enabled you gain the ability to create new template parts.

The reason behind this suggestion is that editors already are able to create new template parts from the template editor as it is in Core today. And if they have the Template Editor enabled they are able to use template parts that they manually created. This is not the case if the template editor is not enabled.

Mamaduka replied:
@mtias, what do you think about this suggestion?

fabiankaegy commented on packages/edit-site/src/components/template-part-converter/convert-to-template-part.js:
```diff
@@ -22,13 +22,25 @@ import { store as noticesStore } from '@wordpress/notices';
 	const { saveEntityRecord } = useDispatch( coreStore );
 	const { createSuccessNotice } = useDispatch( noticesStore );
 
+	const { canCreate } = useSelect( ( select ) => {
```
See https://github.com/WordPress/gutenberg/pull/42729/files#r935247789

richtabor commented:
> I also want to mention this issue, which also touches on the problem of missing alignments in template parts focus mode - https://github.com/WordPress/gutenberg/issues/34010. Cc. @richtabor.

I'm not quite following, testing wise if I pull up a standard block theme should I be able to access top-level alignments for template parts?

Mamaduka commented:
@richtabor, I think if we enabled alignments in template parts focus mode, we should also do it for block themes.

richtabor commented:
> @richtabor, I think if we enabled alignments in template parts focus mode, we should also do it for block themes.

Yes, agreed 🔥

jasmussen commented:
A couple of quick general thoughts on the flow:

* In classic themes, the appearance menu looks way different, with many more options
* Having a “Template Parts” item there which opens the block editor and allows limited editing, that’s cool!
* Using the isolated mode for it, as they are template parts, is also cool.

There is some potential for user confusion: 

* A user of only classic themes switches to a new classic theme with template part support. In using the new feature, they might be confused by features which seem like they should work, but won't in this context, due to limitations that aren’t necessarily communicated.
* Or: a user of mainly block themes switches to a classic theme with template part support and sees this option instead of the site editor, also being confused by new limitations.

In both cases, there may not be a ton we can do, but it might also not be a big problem in practice. If it turns out to be, we could follow up with a minimal welcome-modal that explains the feature, just like the welcome modal for the site editor does.

annezazu commented:
Great thoughts there @jasmussen I'd love to see them shared on this main issue just so it doesn't get lost in the shuffle of this initial PR (if you don't mind): https://github.com/WordPress/gutenberg/issues/37943

Mamaduka committed:
Make sure supportsTemplatePartsMode is false for block themes

Mamaduka committed:
pinking shears

Mamaduka commented:
I just pushed the following changes:
* Rebased the branch so that some recent fixes like #43408 are also available here.
* Updated how this new setting is set so it can't be accidentally enabled for Block Themes by plugins.

I believe this PR is ready for review.

scruffian commented:
Thanks for working on this. When I go to "Template Parts" I am shown the "Site Editor" option, which then goes to a broken screen. Is it possible to hide "Site Editor" and "Templates" in these cases?
<img width="1439" alt="Screenshot 2022-08-23 at 09 03 11" src="https://user-images.githubusercontent.com/275961/186105233-5e26e17e-04c0-4653-be01-6530caec5a55.png">

Mamaduka commented:
That's odd. Both menu items you mention are hidden. Here's how Editor looks for me locally.

Can you re-build the plugin? It might be loading JS from the trunk.

![CleanShot 2022-08-23 at 12 10 45](https://user-images.githubusercontent.com/240569/186107181-1578b6a4-049a-4ff7-85be-79426d0a84e2.png)

scruffian commented:
You're right, it seems to be fine now :)

paaljoachim commented:
Thank you very much George for working on this feature! It is very much appreciated!

Downloaded block-fragments-main.zip. Uploaded and activated. 
Local site has Twenty Twenty One preinstalled.
Downloaded the Gutenberg PR build.

I see this:
![Screenshot 2022-09-02 at 15 57 55](https://user-images.githubusercontent.com/5323259/188165316-fec30203-706a-475d-8368-2165ee3a2179.jpg)

Here I expected to be able to create a new template part. Similar to what we see in FSE themes. I also noticed that in FSE themes we have the 3 dot contextual menu to where I can delete a template part.
![Screenshot 2022-09-02 at 16 15 28](https://user-images.githubusercontent.com/5323259/188168636-f28e1708-6996-4621-bd3e-e3f6e2bbb383.jpg)


Editing the current header minimal.
Clicking the title I see an Area drop down. Which I find very useful.
![Screenshot 2022-09-02 at 16 03 27](https://user-images.githubusercontent.com/5323259/188166353-221efe7a-53f2-47c9-acda-a4079c403627.jpg) This looks pretty similar to the FSE version.
I do see General is active in the drop down. Here it should be Header not General because I am editing a Header area.
I save the change and click the back button in the browser. I thought earlier there was a back button just below the W logo to go back to the Template Parts screen.... I found it a bit strange that I had to click the Browsers back button...

I am back in the Template Parts screen. There is now a 3 dot beside the header-minimal to where I can clear customizations.
(Can still not delete the template part though.)
![Screenshot 2022-09-02 at 16 08 47](https://user-images.githubusercontent.com/5323259/188167345-d2562f7a-031b-406c-bfeb-edf159a4abb3.jpg)

--------

Ok we have these very useful template parts. How do I easily use these for a specific page or post, or multiple pages/posts?

fabiankaegy reviewed (approved):
This should not be used as final approval for merging since I am biased. But still I want to indicate that from my perspective and my testing this is ready to go :)

paaljoachim commented:
What is the reason to not including an Add New button to where the user can add a new template?
What about deleting a template?

fabiankaegy commented:
@paaljoachim From my perspective in classic themes there is no way for users to use a template part if the theme doesn't already calls it from one of it's php templates. Therefore it would be confusing to be able to to create or delete the template part without being able to use it anywhere. 

That is why I personally think we should not expose the ability to create or delete. 

This only changes if the theme supports the template editor (by adding a `theme.json` file) at that point they can use the template part block in the template editor to use the template parts anywhere they like. 

So to me it would make sense to only enable these options if the template editor is enabled.

paaljoachim commented:
Thanks @fabiankaegy 
Based on your comment it seems this PR is ready to go. 
I would suggest a last look from @jasmussen @jameskoster as well as another developer @talldan

scruffian commented:
I rebased this so I could test it again.

scruffian reviewed (approved):
LGTM

jasmussen commented:
It's an important feature, and it doesn't preclude future enhancements.

jasmussen commented:
Let's follow up with some documentation on how this is to be used.

Mamaduka commented:
Thanks, everyone, for testing and reviewing. Let's merge this and work on follow-ups.

fabiankaegy commented:
## Dev Note Draft:


---

# Allow usage of block based "template parts" in traditional themes

Starting in WordPress 6.1 even traditional WordPress themes can adopt the usage of block based template parts in their themes. In order to enable this feature a theme needs to specify the `block-template-parts` theme support. With that theme support added a theme can now add block based template parts by placing html files containing the block template into the `parts` directory inside the root of the theme. 

```php 
function add_block_template_part_support() {
    add_theme_support( 'block-template-parts' );
}

add_action( 'after_theme_setup', 'add_block_template_part_support' );
```

These block based template parts can now be used in the traditional PHP templates by using the `block_template_part` function. This function returns the markup of the template part which then needs to get output. 

### Example: 

If a theme wants to make the site footer editable using blocks, the theme author can create a new file called `footer.html` inside the `my-theme/parts/` directory. This file should contain the default footer block markup.

```html
<!-- wp:group {"layout":{"inherit":true}} -->
<div class="wp-block-group">
	<!-- wp:group {"style":{"spacing":{"padding":{"top":"80px","bottom":"30px"}}}} -->
	<div class="wp-block-group" style="padding-top:80px;padding-bottom:30px">
		<!-- wp:paragraph {"align":"center"} -->
		<p class="has-text-align-center">Proudly Powered by <a href="https://wordpress.org" rel="nofollow">WordPress</a></p>
		<!-- /wp:paragraph -->
	</div>
	<!-- /wp:group -->
</div>
<!-- /wp:group -->
```

In order to actually use this template part the theme author then needs to call the `block_template_part` function and pass the name of the template part as the first and only parameter.

```php
<?php block_template_part( 'footer' ); ?>
```

There are also two special helper functions in WordPress. One to retrieve the `header` and one to retrieve the `footer` template parts directly without needing to pass the name manually. `block_header_area` and `block_footer_area`. 

_Notes: In contrast to the template part editor that is included in the Site Editor for block based themes, editors cannot create new template parts, not delete existing ones. This is because there is no way to actually use a newly created template part without changing the code of the theme itself._


---

@Mamaduka would love to get your feedback on this draft :) 

Ideally I would love to use an example of something that isn't the site footer or the site header because I think it makes it slightly confusing with the special helper functions that are available. But even with a different example I think the overall structure could stay the same. 

--- 

Note: We moved to editing of the DEV Note to a google doc that can be found here: https://docs.google.com/document/d/12u_TMKdsRDzdgrqXN2eg3gImaMP3vFMQH4ALUA-1F3A/

ALJ commented:
@fabiankaegy thank you for proposing and advocating for this feature over the last few months!

> Ideally I would love to use an example of something that isn't the site footer or the site header because I think it makes it slightly confusing with the special helper functions that are available.

If you're looking for more ideas: I will make use of the feature for complex navigation menus and other site-wide modals necessitating specific editor-designed layouts.

mtias commented:
@Mamaduka thanks for the perseverance, this is going to be a key part of the puzzle.

bph commented:
@fabiankaegy @Mamaduka Thanks so much for the Dev Note Draft link. When you are finished drafting, please add it as Draft to the Make core blog (see #42961)

AdelDima commented:
This is great news, I'll test it and implement this on our classic theme for sure.

jasmussen commented:
@Mamaduka Nice work on this one. In testing this today, I noticed the Navigation button is present in this mode:

<img width="1448" alt="Screenshot 2022-09-28 at 09 28 33" src="https://user-images.githubusercontent.com/1204802/192716462-4f2edcf6-cf1d-4383-9fb0-32bcf8810dc3.png">

Is this intentional or not? I don't see it as a big problem, it just shows "no navigation menus", but let me know if this needs a ticket or not.

Mamaduka commented:
@jasmussen, not intentional. I left the sidebars as they were. Also, the template parts can use the Navigation block, so I think that can be useful.

What do you think?

jasmussen commented:
I think let's keep it as is for now and just have an ear to the ground on any confusion!

Mamaduka commented:
I am removing the "Needs Dev Note" label. Thank you @fabiankaegy for writing and publishing the post on the make blog 🙇

bph commented:
thanks @Mamaduka I think I put a check mark next to the PR in the DEV Note tracking issue . Don't remove the label, unless you also add `has dev note`🙂

ethanclevenger91 commented:
I'm giving this a spin and it's possible I'm doing something wrong, but it seems like any `contentSize` configured in `theme.json` doesn't get applied to parts included this way in classic themes. Am I missing something, or is that expected behavior?

Mamaduka commented:
Hi, @ethanclevenger91

Do you mean on the front end or in Editor?

ALJ commented:
I see the same behavior @ethanclevenger91 mentioned in both the editor and the front end.

I (maybe wrongly) assumed that was intentional behavior when editing Template Parts to ignore `contentSize` on immediate child block of the Part.

To get the `theme.json` -> `contentSize` to apply within a Template Part, I use a `core.group` block as a direct child of the Part and I set its `Inner blocks use content width` setting to `true`.

ethanclevenger91 commented:
@Mamaduka Yeah same as @ALJ - neither frontend nor editor respect `contentSize`

oandregal commented on lib/compat/wordpress-6.1/template-parts-screen.php:
```diff
@@ -0,0 +1,208 @@
+		}
+	}
+
+	$active_global_styles_id = WP_Theme_JSON_Resolver::get_user_global_styles_post_id();
```
@Mamaduka @scruffian @fabiankaegy @Soean We should use the `WP_Theme_JSON_Resolver_Gutenberg` class instead of the core one. Fix for this at https://github.com/WordPress/gutenberg/pull/49198
